src/v/cluster/tx_gateway_frontend.cc

```diff
@@ -46,6 +70,892 @@ tx_gateway_frontend::tx_gateway_frontend(
+          }
+
+          auto tx = maybe_tx.value();
+          if (tx.status != tm_transaction::tx_status::ongoing) {
+              return ss::make_ready_future<kafka::end_txn_response_data>(
+                kafka::end_txn_response_data{
+                  .error_code = kafka::error_code::unknown_server_error});
+          }
+
+          model::producer_identity pid{
+            .id = request.producer_id, .epoch = request.producer_epoch};
+          if (tx.pid != pid) {
+              if (tx.pid.id == pid.id && tx.pid.epoch > pid.epoch) {
+                  return ss::make_ready_future<kafka::end_txn_response_data>(
+                    kafka::end_txn_response_data{
+                      .error_code = kafka::error_code::invalid_producer_epoch});
+              }
+
+              return ss::make_ready_future<kafka::end_txn_response_data>(
+                kafka::end_txn_response_data{
+                  .error_code = kafka::error_code::unknown_server_error});
+          }
+
+          auto outcome = ss::make_lw_shared<ss::promise<tx_errc>>();
+          if (request.committed) {
+              (void)self.commit_tm_tx(stm, tx, timeout, outcome);
+          } else {
+              (void)self.abort_tm_tx(stm, tx, timeout, outcome);
+          }
+
+          return outcome->get_future().then([](tx_errc ec) {
+              if (ec == tx_errc::none) {
+                  return kafka::end_txn_response_data{
+                    .error_code = kafka::error_code::none};
+              }
+
+              return kafka::end_txn_response_data{
+                .error_code = kafka::error_code::unknown_server_error};
+          });
+      });
+}
+
+ss::future<checked<cluster::tm_transaction, tx_errc>>
+tx_gateway_frontend::abort_tm_tx(
+  ss::shared_ptr<cluster::tm_stm>& stm,
+  cluster::tm_transaction tx,
+  model::timeout_clock::duration timeout,
+  ss::lw_shared_ptr<ss::promise<tx_errc>> outcome) {
+    return stm->get_end_lock(tx.id)->with([this, &stm, timeout, tx, outcome]() {
+        return do_abort_tm_tx(stm, tx, timeout, outcome);
+    });
+}
+
+ss::future<checked<cluster::tm_transaction, tx_errc>>
+tx_gateway_frontend::do_abort_tm_tx(
+  ss::shared_ptr<cluster::tm_stm>& stm,
+  cluster::tm_transaction tx,
+  model::timeout_clock::duration timeout,
+  ss::lw_shared_ptr<ss::promise<tx_errc>> outcome) {
+    auto changed_tx = co_await stm->try_change_status(
+      tx.id, tx.etag, cluster::tm_transaction::tx_status::aborting);
+    if (!changed_tx.has_value()) {
+        outcome->set_value(tx_errc::timeout);
+        co_return checked<cluster::tm_transaction, tx_errc>(tx_errc::timeout);
+    }
+    outcome->set_value(tx_errc::none);
+
+    tx = changed_tx.value();
+    std::vector<ss::future<abort_tx_reply>> pfs;
+    for (auto rm : tx.partitions) {
+        pfs.push_back(
+          _rm_partition_frontend.local().abort_tx(rm.ntp, tx.pid, timeout));
+    }
+    std::vector<ss::future<abort_group_tx_reply>> gfs;
+    for (auto group : tx.groups) {
```

Comment:
this makes two copies of the group: one for the iteration object then one passed to abort_group_tx